File: src/org/bireme/dia/analysis/SynonymFilter.java
package org.bireme.dia.analysis;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Stack;

import org.apache.lucene.analysis.TokenFilter;
import org.apache.lucene.analysis.TokenStream;
import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
import org.apache.lucene.analysis.tokenattributes.PositionIncrementAttribute;
import org.apache.lucene.util.AttributeSource;

public class SynonymFilter extends TokenFilter {
    public static final String TOKEN_TYPE_SYNONYM = "SYNONYM";
    public static final String TOKEN_TYPE_WORD = "WORD";

    private final Stack<String> synonymStack;
    private final SynonymEngine engine;
    private final boolean addWords;
    private final boolean processOnlyPrecodTerms;
    private final CharTermAttribute termAtt;
    private final PositionIncrementAttribute posIncrAtt;

    private AttributeSource.State current;

    public SynonymFilter(final TokenStream in,
                         final SynonymEngine engine,
                         final boolean words,
                         final boolean precod) {
        super(in);
        synonymStack = new Stack<String>();
        this.engine = engine;
        this.addWords = words;
        this.processOnlyPrecodTerms = precod;

        this.termAtt = addAttribute(CharTermAttribute.class);
        this.posIncrAtt = addAttribute(PositionIncrementAttribute.class);
    }

    @Override
    public final boolean incrementToken() throws IOException {
    if (! synonymStack.isEmpty()) {
            final String syn = synonymStack.pop();
            restoreState(current);

            termAtt.setEmpty();
            termAtt.append(syn.trim());
            //posIncrAtt.setPositionIncrement(0);
            return true;
    }

    if (!input.incrementToken()) {
            return false;
        }

    if (addAliasesToStack()) {
            current = captureState();
    }

    return true;
    }

    private boolean addAliasesToStack() throws IOException {
        String currentTerm = termAtt.toString();
        //check if only process precod (^d28631) terms
        if (processOnlyPrecodTerms && !currentTerm.startsWith("^d")) {
            return false;
        }

        final String[] synonyms = engine.getSynonyms(currentTerm);
        if (synonyms == null) {
            return false;
        }

        for (String synonym : synonyms) {
            // generate separate tokens (search keys) word by word or term
            if (addWords) {
                // first add the original term to the index
                synonymStack.push(synonym);

                final String synonym_normalized = synonym.replaceAll("/", " ");
                final String[] termWords = synonym_normalized.split(" ");

                if (termWords.length > 1){
                    // inverte a ordem de percorrer os termos devido ao stack sempre adicionar no inicio da pilha
                    for (int i = termWords.length-1; i >= 0; i--) {
                        if (!termWords[i].isEmpty()) {
                            synonymStack.push(termWords[i]);
                            if (termWords[i].contains("-")) {
                                // split keys like omega-3 in 2 tokens (omega 3)
                                final List<String> splited_by_hyphen =
                                        Arrays.asList(termWords[i].split("-"));
                                synonymStack.addAll(splited_by_hyphen);
                            }
                        }
                    }
                }
            } else {
                synonymStack.push(synonym);
            }
        }

        return true;
    }
}

File: src/main/scala/org/bireme/dia/filters/SynonymFilter.scala
package org.bireme.dia.filters

import org.apache.lucene.analysis.tokenattributes.CharTermAttribute
import org.apache.lucene.analysis.{TokenFilter, TokenStream}
import org.apache.lucene.util.AttributeSource
import org.bireme.dia.synonyms.DeCSSynonyms
import org.bireme.dia.tools.Tools

import scala.collection.mutable
import scala.compiletime.uninitialized

class SynonymFilter(in: TokenStream,
                    syns: DeCSSynonyms) extends TokenFilter(in):   // remove \W characters from prefix and suffix of tokens

  private val synonymSet: mutable.Set[String] = new mutable.TreeSet[String]()
  private val termAtt = addAttribute(classOf[CharTermAttribute])
  private var current: AttributeSource.State = uninitialized

  override def close(): Unit =
    //syns.close()
    super.close()

  override def incrementToken(): Boolean =
    if synonymSet.isEmpty then
      if in.incrementToken() then
        val currentTerm: String = termAtt.toString.trim
        val currentTerm2: String = if syns.conf.toLowerCase then currentTerm.toLowerCase else currentTerm
        val currentTerm3: String = if syns.conf.fixPrefixSuffix then Tools.fixPrefixSuffix(currentTerm2) else currentTerm2

        if currentTerm3.isEmpty then incrementToken()
        else
          synonymSet.add(currentTerm3)
          addAliasesToSet(currentTerm)
          val syn: String = synonymSet.head.trim
          synonymSet.remove(syn)
          //restoreState(current)
          termAtt.setEmpty()
          termAtt.append(syn)
          true
      else false
    else
      val syn: String = synonymSet.head
      //val contain: Boolean = synonymSet.contains(syn)
      synonymSet.remove(syn)
      //restoreState(current)
      termAtt.setEmpty()
      termAtt.append(syn)
      //posIncrAtt.setPositionIncrement(0)
      true

  private def addAliasesToSet(currentTerm: String): Boolean =
    val synonyms: Set[String] = syns.getSynonyms(currentTerm)

    synonyms.foreach(elem => synonymSet.add(elem.trim))
    synonyms.isEmpty